### PR TITLE
Assignment 2 - Paulson Ray Thompson

### PR DESCRIPTION
**Pull request by Purplepigs0129**
Here it is hopefully done

### CODE REVIEW

**Comment by Purplepigs0129** [comment]
Originally had a remove second to work on my machine if the test fails on yours just add it in appropriately using .addTime on fileloggertests.cs

**Comment by BenjaminMichaelis** [comment]
Ignore the one specific failure on https://github.com/IntelliTect-Samples/EWU-CSCD371-2024-Winter/actions/runs/7621660789/job/20758379640?pr=62, not sure why a second action triggered.

**Comment by Tvor369** [comment]
reviewing

File: Logger.Tests/Logger.Tests.csproj [hunk 1]
```diff
@@ -1,9 +1,14 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
 	<PropertyGroup>
-		<TargetFramework>net7.0</TargetFramework>
+		<TargetFramework>net8.0</TargetFramework>
 		<IsPackable>false</IsPackable>
-		<LangVersion>11.0</LangVersion>
+    <LangVersion>11.0</LangVersion>
```
Comment (Tvor369) [inline Logger.Tests/Logger.Tests.csproj:6]: Latest LangVersion of C# is 12.0 not 11.0

File: Logger/BaseLogger.cs [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 public abstract class BaseLogger
 {
-    public abstract void Log(LogLevel logLevel, string message);
+    public abstract void Log(LogLevel LogLevel, string Message);
```
Comment (Tvor369) [inline Logger/BaseLogger.cs:5]: Don't we use camelCase for variable names? Maybe I'm mistaken here.
```suggestion
    public abstract void Log(LogLevel logLevel, string message);
```

File: Logger/LogFactory.cs [hunk 3]
```diff
@@ -2,9 +2,25 @@
...
+    public BaseLogger? CreateLogger (string className)
+    {
+        if (className == nameof(FileLogger))
+        {
+            FileLogger FileLogger = new(_loggerPath!) { ClassName = className };
+            return FileLogger;
+        } else
+        {
+            return null;
+        }
```
Comment (Tvor369) [inline Logger/LogFactory.cs:16]: Definitely not wrong but you could consider removing the else and just have return null by itself,  if the if statement didn't trigger then we wouldn't get that earlier return FileLogger and would instead be executing whatever is after the if statement anyway.
```suggestion
        }
        
        return null;
        
```

File: Logger/Logger.csproj [hunk 4]
```diff
@@ -1,8 +1,13 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
-    <LangVersion>11.0</LangVersion>
+    <TargetFramework>net7.0</TargetFramework>
```
Comment (Tvor369) [inline Logger/Logger.csproj:4]: Should be net8.0
```suggestion
    <TargetFramework>net8.0</TargetFramework>
```

File: Logger.Tests/LogFactoryTests.cs [hunk 5]
```diff
@@ -1,9 +1,27 @@
...
+        _logFactory = new();
+        _logFactory.ConfigureFileLogger("Test.txt");
+        _logger = _logFactory.CreateLogger("Test");
+    }
+
+    [TestMethod]
+    public void ConfigureFileLogger_GoodFilePath_Successful()
+    {
+        Assert.AreEqual("Test.txt",_logFactory!.GetFileName());
+    }
```
Comment (Tvor369) [inline Logger.Tests/LogFactoryTests.cs:26]: I think you should add a test for the case when CreateLogger returns null when the filePath isn't set with ConfigureFileLogger.

**Comment by Tvor369** [comment]
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not been configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested.

Overall good job, I can tell you tried to set the LangVersion and TargetFramework correctly as the project file for your logger.tests had the right TargetFramwork but not LangVersion, and your logger project file had the opposite issue. The other thing is I think you need to add a test to LogFactory for returning null when no filePath is set.

File: Logger/Logger.csproj (hunk 4 above)
Comment (Purplepigs0129) [reply Logger/Logger.csproj:4]: Yes it should originally had these set to the proper ones might have gotten mixed up in the process of trying to get the checks to pass.

File: Logger/LogFactory.cs (hunk 3 above)
Comment (Purplepigs0129) [reply Logger/LogFactory.cs:16]: Ahh thanks for the catch and explanation will fix

File: Logger/BaseLogger.cs (hunk 2 above)
Comment (Purplepigs0129) [reply Logger/BaseLogger.cs:5]: Yes a little mistake I found to late on Monday to fix had it in my plans thanks for the catch

File: Logger.Tests/Logger.Tests.csproj (hunk 1 above)
Comment (Purplepigs0129) [reply Logger.Tests/Logger.Tests.csproj:6]: correct got mixed up int eh checks earlier will be fixed

File: Logger.Tests/LogFactoryTests.cs (hunk 5 above)
Comment (Purplepigs0129) [reply Logger.Tests/LogFactoryTests.cs:26]: Will do just ran out of time on monday

File: Logger/BaseLoggerMixins.cs [hunk 6]
```diff
@@ -1,6 +1,73 @@
-﻿namespace Logger;
+﻿using System;
+using System.Runtime.CompilerServices;
+
+namespace Logger;
 
 public static class BaseLoggerMixins
 {
+    public static void Error(this BaseLogger? logger, string message, params object[] args)
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:8]: Why is logger nullable? I expect to call this message, a logger is needed.

File: Logger.Tests/FileLoggerTests.cs [hunk 7]
```diff
@@ -1,9 +1,44 @@
...
+        string message = "Hello Everybody!";
+       
+        if (File.Exists(_Path))
+        {
+            File.Delete(_Path);
+        }
+
+        FileStream MyFile = File.Create(_Path);
+        MyFile.Close();
+        _fileLogger!.ClassName = "Test";
```
Comment (BenjaminMichaelis) [inline Logger.Tests/FileLoggerTests.cs:32]: Why are you using a `!` here instead of checking if the logger is in fact not null?

File: Logger.Tests/Logger.Tests.csproj [hunk 8]
```diff
@@ -1,9 +1,14 @@
...
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <EnableNETAnalyzers>true</EnableNETAnalyzers>
+    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
+    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
+    <Nullable>enable</Nullable>
```
Comment (BenjaminMichaelis) [inline Logger.Tests/Logger.Tests.csproj:11]:
```suggestion
		<LangVersion>11.0</LangVersion>
		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
		<EnableNETAnalyzers>true</EnableNETAnalyzers>
		<CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
		<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
		<Nullable>enable</Nullable>
```
nit: spacing

File: Logger.Tests/Logger.Tests.csproj (hunk 1 above)
Comment (BenjaminMichaelis) [inline Logger.Tests/Logger.Tests.csproj:6]: Latest is 12.0

File: Logger/BaseLogger.cs [hunk 9]
```diff
@@ -3,5 +3,7 @@
 public abstract class BaseLogger
 {
     public abstract void Log(LogLevel logLevel, string message);
+
+    public string? ClassName { get; set; }
```
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:7]: Why is this nullable?

File: Logger/FileLogger.cs [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+using System.Text;
+
+namespace Logger;
+
+public class FileLogger : BaseLogger
+{
+    private readonly string _Path;
+    public FileLogger(string filePath)
+    {
+        this._Path = filePath;
```
Comment (BenjaminMichaelis) [inline Logger/FileLogger.cs:13]:
```suggestion
        _Path = filePath;
```
why is this used

File: Logger/FileLogger.cs [hunk 11]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+namespace Logger;
+
+public class FileLogger : BaseLogger
+{
+    private readonly string _Path;
```
Comment (BenjaminMichaelis) [inline Logger/FileLogger.cs:10]: This should be a property, not a field

File: Logger/LogFactory.cs [hunk 12]
```diff
@@ -2,9 +2,25 @@
 
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
-    {
+    private string? _loggerPath;
 
+    public BaseLogger? CreateLogger (string className)
+    {
+        if (className == nameof(FileLogger))
```
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:9]: This is weird.... How is this supposed to take a class name with this check?

File: Logger/LogFactory.cs [hunk 13]
```diff
@@ -2,9 +2,25 @@
...
+            FileLogger FileLogger = new(_loggerPath!) { ClassName = className };
+            return FileLogger;
+        }
+        
         return null;
+        
+    }
+    public void ConfigureFileLogger(string FilePath)
+    {
+        this._loggerPath = FilePath;
```
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:20]: `this` should not be needed here
